dv/verilator/cs_registers/cpp/reg_model.cc

```diff
@@ -0,0 +1,226 @@
+  uint32_t pmp_cfg_addr = 0x3A0 + (pmp_region / 4);
+  // Form the address of the CFG registerfor the next region
+  // For region 15, this will point to a non-existant register, which is fine
+  uint32_t pmp_cfg_plus1_addr = 0x3A0 + ((pmp_region + 1) / 4);
+  uint32_t cfg_value = 0;
+  uint32_t cfg_plus1_value = 0;
+  // Find and read the two CFG registers
+  for (std::vector<BaseRegister *>::iterator it = map_pointer_->begin();
+       it != map_pointer_->end(); ++it) {
+    if ((*it)->MatchAddr(pmp_cfg_addr)) {
+      cfg_value = (*it)->RegisterRead();
+    }
+    if ((*it)->MatchAddr(pmp_cfg_plus1_addr)) {
+      cfg_plus1_value = (*it)->RegisterRead();
+    }
+  }
+  // Shift to the relevant bits in the CFG registers
+  cfg_value >>= ((pmp_region & 0x3) * 8);
+  cfg_plus1_value >>= (((pmp_region + 1) & 0x3) * 8);
+  // Locked if the lock bit is set, or the next region is TOR
+  if ((cfg_value & 0x80) || ((cfg_plus1_value & 0x18) == 0x8)) {
+    return 0xFFFFFFFF;
+  } else {
+    return 0;
+  }
+}
+
+uint32_t NonImpRegister::RegisterRead() { return 0; }
+
+uint32_t NonImpRegister::RegisterWrite(uint32_t newval) { return 0; }
+
+uint32_t NonImpRegister::RegisterSet(uint32_t newval) { return 0; }
+
+uint32_t NonImpRegister::RegisterClear(uint32_t newval) { return 0; }
+
+RegisterModel::RegisterModel(RegisterInterface *reg_signals,
+                             EnvInterface *env_signals)
+    : reg_signals_(reg_signals), env_signals_(env_signals) {
+
+  for (int i = 0; i < 4; i++) {
+    uint32_t reg_addr = 0x3A0 + i;
+    if (i < (env_signals_->PMP_NUM_REGIONS / 4)) {
+      register_map_.push_back(new PmpCfgRegister(reg_addr, &register_map_));
+    } else {
+      register_map_.push_back(new NonImpRegister(reg_addr, &register_map_));
+    }
+  }
+  for (int i = 0; i < 16; i++) {
+    uint32_t reg_addr = 0x3B0 + i;
+    if (i < env_signals_->PMP_NUM_REGIONS) {
+      register_map_.push_back(new PmpAddrRegister(reg_addr, &register_map_));
+    } else {
+      register_map_.push_back(new NonImpRegister(reg_addr, &register_map_));
+    }
+  }
+}
+
+RegisterModel::~RegisterModel() {
+  for (std::vector<BaseRegister *>::iterator it = register_map_.begin();
+       it != register_map_.end(); ++it) {
+    delete *it;
+  }
+  register_map_.clear();
+}
+
+void RegisterModel::RegisterReset() {
+  for (std::vector<BaseRegister *>::iterator it = register_map_.begin();
+       it != register_map_.end(); ++it) {
+    (*it)->RegisterReset();
+  }
+}
+
+int RegisterModel::RegisterHandler() {
+  if (!env_signals_->rst_n) {
+    RegisterReset();
+    return 0;
+  }
+  // TODO add machine mode permissions to registers
+  if (reg_signals_->reg_access) {
+    uint32_t read_val;
+    bool matched = 0;
+    for (std::vector<BaseRegister *>::iterator it = register_map_.begin();
+         it != register_map_.end(); ++it) {
+      if ((*it)->MatchAddr(reg_signals_->reg_addr)) {
+        matched = 1;
+        switch (reg_signals_->reg_op) {
+        case 0: {
```

Comment:
In the interests of clarity, this should be an enum (potentially an enum class) not an integer.

Reply:
enum added